influxql/call_iterator.go

```diff
@@ -362,18 +360,30 @@ func booleanLastReduce(prev, curr *BooleanPoint) (int64, bool, []interface{}) {
 func NewDistinctIterator(input Iterator, opt IteratorOptions) (Iterator, error) {
 	switch input := input.(type) {
 	case FloatIterator:
-		return &floatReduceSliceIterator{input: newBufFloatIterator(input), opt: opt, fn: floatDistinctReduceSlice}, nil
+		createFn := func() (FloatPointSliceAggregator, FloatPointSliceEmitter) {
+			fn := NewFloatSliceFuncReducer(FloatDistinctReduceSlice)
+			return fn, fn
+		}
+		return &floatReduceSliceIterator{input: newBufFloatIterator(input), opt: opt, create: createFn}, nil
 	case IntegerIterator:
-		return &integerReduceSliceIterator{input: newBufIntegerIterator(input), opt: opt, fn: integerDistinctReduceSlice}, nil
+		createFn := func() (IntegerPointSliceAggregator, IntegerPointSliceEmitter) {
+			fn := NewIntegerSliceFuncReducer(IntegerDistinctReduceSlice)
+			return fn, fn
+		}
+		return &integerReduceSliceIterator{input: newBufIntegerIterator(input), opt: opt, create: createFn}, nil
 	case StringIterator:
-		return &stringReduceSliceIterator{input: newBufStringIterator(input), opt: opt, fn: stringDistinctReduceSlice}, nil
+		createFn := func() (StringPointSliceAggregator, StringPointSliceEmitter) {
+			fn := NewStringSliceFuncReducer(StringDistinctReduceSlice)
+			return fn, fn
+		}
+		return &stringReduceSliceIterator{input: newBufStringIterator(input), opt: opt, create: createFn}, nil
 	default:
 		return nil, fmt.Errorf("unsupported distinct iterator type: %T", input)
 	}
 }
 
-// floatDistinctReduceSlice returns the distinct value within a window.
-func floatDistinctReduceSlice(a []FloatPoint, opt *reduceOptions) []FloatPoint {
+// FloatDistinctReduceSlice returns the distinct value within a window.
+func FloatDistinctReduceSlice(a []FloatPoint, opt *ReduceOptions) []FloatPoint {
```

Comment:
It's mostly because I think it makes a better implementation. Instead of having to pass the start time to every function and have potentially unique behavior of how it sets the time for aggregates, the caller determines how the time gets set. The reducers already only have partial knowledge of the point they're dealing with. The iterator that was created for reduce operations checks to see if the time is set to `ZeroTime` and overwrites it with the interval time otherwise it keeps the time as it is and a later iterator modifies the time field to ensure uniform handling in the system. There are also some aggregates that do something different depending on if an interval was set or not. For example, the `first()` call will return the time of the point if there is no interval and will return the window time if there is an interval. The iterator has that knowledge by checking the `GROUP BY` in the `IteratorOptions`. `ReduceOptions` has to guess using the start time and the end time which isn't a reliable method. @benbjohnson didn't want `nil` being passed when there was no interval, so I passed on that. I think he's right that `nil` shouldn't be passed into the function when there is no interval.

At that point, I couldn't think of any useful reason why the reducers needed to know the time.

So I wanted the reducers to just focus on reducing and none of them seemed to care about the time period. Separation of concerns and I only have to implement the logic for setting the time in one location instead of every location.

It also means that I get to remove and never expose `ReduceOptions` which leads to less code that has to be maintained. In my mind, `ReduceOptions` isn't being used since it's almost immediately overwritten by other parts of the system and doesn't seem to add to the implementation other than an extra parameter that gets ignored.

#5900 also contributes to this section of the code no longer being useful since I needed a higher level of the query engine to normalize times anyway.